testing-unit-cs-top-level-program/README.md

```diff
@@ -0,0 +1,78 @@
+# C# Unit Testing with Top-Level Statements
+
+In this example project, we examine how to unit test C# Pulumi programs that are using top-level statements with dotnet SDK v6. These programs look like this
+
+```csharp
+using Pulumi;
+
+return await Deployment.RunAsync(() => 
+{
+    // Create resources here
+});
+```
+
+In order to unit test this piece of code, we have to do a bit of rewriting such that we don't run the `Deployment.RunAsync` function because it expects the program is being executed from the Pulumi CLI. 
+
+Now, we take the lambda that is provided to `Deployment.RunAsync` and move it in a place where we can reference it later from our unit test project. The following snippet shows this
+
+```csharp
+using Pulumi;
+
+return await Deployment.RunAsync(Deploy.Infra);
+
+class Deploy
+{
+    public static Dictionary<string, object?> Infra()
+    {
+        // Create resources here
+    }
+}
+```
+Here, the code that creates resources and returns outputs is inside the `Deploy.Infra` function which we can reference from the test project. 
+
+> The return type of `Deploy.Infra` could also have been any of the following 
+> - `void` when we are not returning outputs
+> - `Task` when we want to `await` async code
+> - `Task<Dictionary<string, object?>>` when we want to `await` async code _and_ return outputs
+
+There is a catch before we can reference `Deploy.Infra` from the unit tests. Since we don't have an explicit namespace here, we need to add this property to the project file
+```xml
+<ItemGroup>
+  <InternalsVisibleTo Include="Tests"/>
+</ItemGroup>
+```
+Now we are ready to write unit tests. From a test project called `Tests`, add a reference to the project that contains the Pulumi program:
+```xml
+<ItemGroup>
+  <ProjectReference Include="..\infra\Infra.csproj" />
+</ItemGroup>
+```
+Now, only for the tests we can create a _test stack_ that uses `Deploy.Infra` as follows
+```csharp
+class TestStack : Stack
+{
+    public TestStack()
+    {
+        Outputs = Deploy.Infra();
+    }
+
+    public Dictionary<string, object?> Outputs { get; set; }
+}
+```
+Then, we can use this test stack to build resources and make assertions about them:
+```csharp
+[Test]
+public async Task StackExportsWebsiteUrl()
+{
+    var options = new TestOptions { IsPreview = false };
+	var resources = await Pulumi.Deployment.TestAsync<TestStack>(new Mocks(), options);
```

Comment:
I wonder if we should have a TestAsync method that takes the same delegates that RunAsnyc takes? Would save the dummy stack class above.

Reply:
Yeah, +1 to that. `TestStack` is a bit ugly. Also, will it work if `Infra` is async?

Reply:
> will it work if Infra is async?

Yes, but the user will have to modify the code a bit:
```cs
Outputs = Deploy.Infra().GetAwaiter().GetResult();
```

> I wonder if we should have a TestAsync method that takes the same delegates that RunAsnyc takes? 

Yes, that's the plan. I also would love it if the following code works
```cs
await Deployment.TestAsync(mocks, options, () => 
{
    return Deployment.RunAsync(() => 
    {
        // Create resources here
    });
});
```
Because then, without changing a user's code when they write
```cs
// Program.cs
return await Deployment.RunAsync(() => 
{
    // Create resources here
});
``` 
We could write a utility function for unit tests that dynamically executes the entry point of the assembly being tested _inside_ a `Deployment.TestAsync` call. No need to rewrite the code to make it unit-testable